File: 高级数据结构与算法/C/第4章 模式匹配/4.2 Sunday/sunday.c
#include <stdio.h>
#include <string.h>

int sunday(char *s, char *p) {
    int sLen = strlen(s);
    int pLen = strlen(p);
    int i = 0;
    int j = 0;
    int result = 0;

    while(i < sLen && j < pLen) {
        if(s[i] == p[j]) {
            i++;
            j++;
            continue;
        }

        int idx = result + pLen;
        if(idx >= sLen) {
            return -1;
        }

        int k = pLen - 1;
        while(k >= 0 && s[idx] != p[k]) {
            k--;
        }

        i = result;
        i += pLen - k;
        result = i;
        j = 0;

        if(result + pLen > sLen) {
            return -1;
        }
    }

    return result;
}

int main() {
    char *s = "a simple example";
    char *p = "ample";
    printf("%d\n", sunday(s, p));
    return 0;
}

File: code/Python/第16章 模式匹配/16.2 Sunday/sunday.py
def sunday(s, p):
    s_len = len(s)
    p_len = len(p)
    i = 0
    j = 0
    result = 0
    
    while i < s_len and j < p_len:
        if s[i] == p[j]:
            i += 1
            j += 1
            continue

        idx = result + p_len
        if idx >= s_len:
            return -1

        k = p_len - 1
        while k >= 0 and s[idx] != p[k]:
            k -= 1

        i = result
        i += p_len - k
        result = i
        j = 0

        if result + p_len > s_len:
            return -1

    return result

def main():
    s = "a simple example"
    p = "ample"
    print(sunday(s, p))

if __name__ == "__main__":
    main()